### PR TITLE
finished nodejs week2 homework

### PR DESCRIPTION
**Pull request by chunmeishui**

### CODE REVIEW

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
+//import
+const express = require("express");
+const router = express.Router();
+const meals = require("./../data/meals.json");
+
+router.get("/", async (req, res) => {
+  let data = meals;
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: try to give a descriptive name instead of data, what data is it holding? so the name can be something relating to that.

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+  if ("date" in req.query ) {
+    try {
+      data = data.filter((meal) =>
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: give a descriptive name here as well. It's a little confusing with ``data = data.filter``, I know you're filtering meals but make it a habit to give some good descriptive names.

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    new Date(meal.createdAt).getTime() > new Date(req.query.date).getTime());
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js:13]: why did you need getTime here? Did you test it out?

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    if (data.length > 0) {
+      res.json(data)
+    }else{
+      res.status(404);
+      res.json("there is no such meal date,Pls check");
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: you can send this as a ``res.send`` since it's not a json object.

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  if ("limit" in req.query) {
+    const mealsNew = meals.map(meal => meal);
+    const limitNo = Number(req.query.limit);
+    data = mealsNew.slice(0, limitNo);
+    if (limitNo <= mealsNew.length && limitNo > 0) {
+            res.json(data);
+          } else {
+            res.status(404).json(`Sorry,there is only ${mealsNew.length} meals left.`);
+          }
+  }
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: Since you have an array already you don't need mealsNew to get the length of the meals
You can get it straight from the meals array. Try it out.

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+//     } else {
+//       res.status(404);
+//       res.json("there is no such meal id,Pls check");
+    
+//     }
+//   } catch (error) {
+//     throw error;
+//   }
+
+// });
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js:154]: if you're not going to use the commented out code later you can remove it. Always clean up unused code.

File: nodejs/week2/homework/meal-sharing/src/backend/api/reservation-router.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
+const express = require("express");
+const router = express.Router();
+const reservations = require("./../data/reservations.json");
+const app = express();
+
+router.get("/:id", (req, rep) => {
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/reservation-router.js:6]: can you make it *res* for response instead?

File: nodejs/week2/homework/meal-sharing/src/backend/api/reservation-router.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    try {
+      const specificReservation = reservations.find(reservation => `${reservation.id}` === req.params.id );
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/api/reservation-router.js:8]: nice descriptive name ``specificReservation`` 👍🏾

File: nodejs/week2/homework/meal-sharing/src/backend/app.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+const reviewRouter = require("./api/review-router");
+
+
+// app.use binds middleware to your application. You can give app.use a path and router. The mini router will take care of all requests with the path
+app.use("/api/meals", mealsRouter);
+app.use("/api/reservations", reservationRouter);
+app.use("/api/reviews", reviewRouter);
+
+
+module.exports = app;
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/app.js:15]: nice 👍🏾

File: nodejs/week2/homework/meal-sharing/src/backend/data/reviews.json [hunk 10]
```diff
@@ -0,0 +1,22 @@
+[
+    {
+      "id": 1,
+      "numberOfStars": 3,
+      "content": "Best meal ever",
+      "mealId": 1,
+      "createdAt": 12345
```
Comment (STEP0HEN) [inline nodejs/week2/homework/meal-sharing/src/backend/data/reviews.json]: I will advice you to use proper date, instead of these numbers. You can use the [js date object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) and send the current date.

File: nodejs/week2/homework/warmup/index.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    if (first && second) {
+        res.send(`multiply result is : ${multiplyResult}`)
+    }
+    if (first) {
+        res.send(`you only type the firstNumber only ${first}`)
+    }
+    if (second) {
+
+        res.send(`you only type the secondNumber only${second}`)
+    }
```
Comment (STEP0HEN) [inline nodejs/week2/homework/warmup/index.js]: Did you test this out? What results do you get? I think you should check for the absence of first and second and send the messages so you'll have something like ``if(!first....){ do something}``. Let me know the outcome.

File: nodejs/week2/homework/warmup/index.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    if (first && second) {
+        res.send(`add result is : ${addResult}`)
+    }
+    if (first) {
+        res.send(`you only type the firstNumber only ${first}`)
+    }
+    if (second) {
+
+        res.send(`you only type the secondNumber only${second}`)
+    }
```
Comment (STEP0HEN) [inline nodejs/week2/homework/warmup/index.js]: same here as below

**Review by STEP0HEN — CHANGES_REQUESTED** [review]
Great job @chunmeishui ! Keep up the good work.

I left a few comments for things to improve on, or change. Let me know if something isn't clear.

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js (hunk 4 above)
Comment (chunmeishui) [reply nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: yes, you are right. Thanks.

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js (hunk 1 above)
Comment (chunmeishui) [reply nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: yes, you are right.Thanks :)

Comment (chunmeishui) [reply nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: I already changed to DataOfMeal. Is it okay?

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js (hunk 2 above)
Comment (chunmeishui) [reply nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: Hi Stepohen, as you can see the data is always changing the value in different If sentences. So I did not give it a different name in different results. data here is ways the data we filter out.

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js (hunk 3 above)
Comment (chunmeishui) [reply nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js:13]: hi Stepohen.I tested this and it works fine. If I only use the new Date it will return the YYYY-MM-DD of this format but ignore the time. that is why I use getTime().

File: nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js (hunk 5 above)
Comment (chunmeishui) [reply nodejs/week2/homework/meal-sharing/src/backend/api/meals-router.js]: here I do not want to change the original meals.  Because I will use slice(). It will change the original one.  That is why  I make the array to the new one and change from the new array.

File: nodejs/week2/homework/meal-sharing/src/backend/data/reviews.json (hunk 10 above)
Comment (chunmeishui) [reply nodejs/week2/homework/meal-sharing/src/backend/data/reviews.json]: sorry for that. It is should be the right time. :)

File: nodejs/week2/homework/warmup/index.js (hunk 11 above)
Comment (chunmeishui) [reply nodejs/week2/homework/warmup/index.js]: if (addResult) {
        res.send(`add result is : ${addResult}`) 
    }else{
        res.send(`check the number you typed`)
    }
how do you think to judge it this way?

Comment (chunmeishui) [reply nodejs/week2/homework/warmup/index.js]: haha, yes you are right I should write like the way you mentioned.

File: nodejs/week2/homework/warmup/index.js (hunk 12 above)
Comment (chunmeishui) [reply nodejs/week2/homework/warmup/index.js]: if (multiplyResult) {
        res.send(`add result is : ${addResult}`) 
    }else{
        res.send(`check the number you typed`)
    }

you are right. Thanks for your valuable feedback. I changed everything.
